### PR TITLE
When a channel goes down, kill off all Pipes.

### PR DESCRIPTION
**Pull request by kohsuke**
See the commit message for more details of this change.

This prevents a hang in channel communication.

### CODE REVIEW

**Comment by kohsuke** [comment]
@reviewbybees

**Comment by rsandell** [comment]
:+1: Assuming the test is just flaky.

File: src/main/java/hudson/remoting/ExportTable.java [hunk 1]
```diff
@@ -338,6 +338,38 @@ Object get(int id) {
...
+    public void abort(Throwable e) {
+        List<Entry<?>> values;
+        synchronized (this) {
+            values = new ArrayList<Entry<?>>(table.values());
+        }
+        for (Entry<?> v : values) {
+            if (v.object instanceof ErrorPropagatingOutputStream) {
+                try {
+                    ((ErrorPropagatingOutputStream)v.object).error(e);
+                } catch (IOException x) {
```
Comment (tfennelly) [inline src/main/java/hudson/remoting/ExportTable.java]: Maybe catch `Exception`? Otherwise propagation stops.

Comment (tfennelly) [inline src/main/java/hudson/remoting/ExportTable.java]: Or `Throwable`.

**Comment by jtnord** [comment]
Checking locally if this fixes `PipeTest` that is reliably broken on a single core machine.

**Comment by jtnord** [comment]
`PipeTest.testReaderCloseWhileWriterIsStillWriting-local-compatibility` is still failing locally (modified test to fail after 3 minutes).  Would you have expected that to be fixed here?

File: src/main/java/hudson/remoting/ExportTable.java [hunk 2]
```diff
@@ -338,6 +338,38 @@ Object get(int id) {
...
 
+    /**
+     * Propagate a channel termination error to all the exported objects.
+     * 
+     * <p>
+     * Exported {@link Pipe}s are vulnerable to infinite blocking
+     * when the channel is lost and the sender side is cut off. The reader
+     * end will not see that the writer has disappeared.
+     */
+    public void abort(Throwable e) {
```
Comment (jglick) [inline src/main/java/hudson/remoting/ExportTable.java]: Should probably be package-private.

**Comment by kohsuke** [comment]
Code review comments addressed, but I still need to look at the test failure, it sounds like.

**Comment by kohsuke** [comment]
The tests pass locally for me, and CI build passed for 2bb1421.

So I believe the failure in `PipeTest.testReaderCloseWhileWriterIsStillWriting-local-compatibility` is real but different issue, and maybe I can work with James at some point to dig deeper if he can reproduce this more easily.

I'm going to merge this PR.

**Pull request merged**